### PR TITLE
 feat(FEC-8661): create youtube engine

### PR DESCRIPTION
**Pull request by OrenMe**

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+let plugins = [
+  new webpack.DefinePlugin({
+    __VERSION__: JSON.stringify(packageData.version),
+    __NAME__: JSON.stringify(packageData.name)
+  })
+];
+
+if (PROD) {
+  // plugins.push(new webpack.optimize.UglifyJsPlugin({sourceMap: true}));
```
Comment (itaykinnrot) [inline webpack.config.js]: ?

File: src/youtube.js [hunk 2]
```diff
@@ -0,0 +1,1092 @@
...
+  _isSeeking: boolean = false;
+
+  _firstPlaying: boolean = true;
+
+  _currentState: number | null = null;
+
+  _playerTracks: Array<Track> = [];
+
+  /**
+   * The html5 class logger.
```
Comment (itaykinnrot) [inline src/youtube.js]: html5?

File: src/youtube.js [hunk 3]
```diff
@@ -0,0 +1,1092 @@
...
+   * @type {any}
+   * @static
+   * @private
+   */
+  static _logger: any = getLogger('Youtube');
+
+  /**
+   * The html5 capabilities handlers.
```
Comment (itaykinnrot) [inline src/youtube.js]: html5?

File: src/youtube.js [hunk 4]
```diff
@@ -0,0 +1,1092 @@
...
+  }
+
+  /**
+   * Select a new video track.
+   * @param {VideoTrack} videoTrack - The video track object to set.
+   * @returns {void}
+   */
+  selectVideoTrack(videoTrack: VideoTrack): void {
+    if (this._api) {
+
```
Comment (itaykinnrot) [inline src/youtube.js]: lint

File: src/youtube.js [hunk 5]
```diff
@@ -0,0 +1,1092 @@
...
+  /**
+   * Enables adaptive bitrate switching according to the media source extension logic.
+   * @function enableAdaptiveBitrate
+   * @returns {void}
+   * @public
+   */
+  enableAdaptiveBitrate(): void {
+    if (this._api) {
+      this._api.setPlaybackQuality('default');
```
Comment (itaykinnrot) [inline src/youtube.js]: string to enum?

File: src/youtube.js [hunk 6]
```diff
@@ -0,0 +1,1092 @@
...
+   * @public
+   * @returns {void}
+   */
+  play(): void {
+    if (this._api) {
+      if (this.currentTime === this.duration) {
+        this.currentTime = 0;
+      }
+      this._api.playVideo();
+      // this.dispatchEvent(new FakeEvent(EventType.PLAY));
```
Comment (itaykinnrot) [inline src/youtube.js]: ?

File: src/youtube.js [hunk 7]
```diff
@@ -0,0 +1,1092 @@
...
+  /**
+   * Loads the YouTube player.
+   * @private
+   * @returns {void}
+   */
+  _loadYouTubePlayer(): void {
+    const loadYouTubePlayer = () => {
+      const config = {
+        playerVars: {
+          width: 1920,
```
Comment (itaykinnrot) [inline src/youtube.js]: hard coded values?

File: src/youtube.js (hunk 5 above)
Comment (OrenMe) [reply src/youtube.js]: string, they don't have enums.
In any case, as we have noticed, the API is just a stub and doesn't work.

File: src/youtube.js (hunk 7 above)
Comment (OrenMe) [reply src/youtube.js]: removed, this seems to be dynamically evaluated anyway

File: src/youtube.js [hunk 8]
```diff
@@ -0,0 +1,1092 @@
...
+          height: 1080,
+          controls: 0,
+          iv_load_policy: 3,
+          disablekb: 1,
+          modestbranding: 1,
+          playsinline: this._config.playback.playsinline ? 1 : 0,
+          rel: 0,
+          fs: 0
+        },
```
Comment (odedhutzler) [inline src/youtube.js]: maybe some kind of default config /  or add costs instead of some numbers?

File: src/youtube.js [hunk 9]
```diff
@@ -0,0 +1,1092 @@
...
+   * @private
+   * @returns {void}
+   */
+  _loadYouTubeIframeAPI(): Promise<> {
+    return new Promise((resolve, reject) => {
+      if (window && window.YT && window.YT.Player) {
+        return resolve();
+      }
+      const tag = Utils.Dom.createElement('script');
+      tag.src = 'https://www.youtube.com/iframe_api';
```
Comment (odedhutzler) [inline src/youtube.js]: const maybe

File: src/youtube.js [hunk 10]
```diff
@@ -0,0 +1,1092 @@
...
+  //   this._eventManager.listen(this._mediaSourceAdapter, CustomEventType.ABR_MODE_CHANGED, (event: FakeEvent) => this.dispatchEvent(event));
+  //   this._eventManager.listen(this._mediaSourceAdapter, CustomEventType.TEXT_CUE_CHANGED, (event: FakeEvent) => this.dispatchEvent(event));
+  //   this._eventManager.listen(this._mediaSourceAdapter, CustomEventType.TRACKS_CHANGED, (event: FakeEvent) => this.dispatchEvent(event));
+  //   this._eventManager.listen(this._mediaSourceAdapter, Html5EventType.ERROR, (event: FakeEvent) => this.dispatchEvent(event));
+  //   this._eventManager.listen(this._mediaSourceAdapter, Html5EventType.TIME_UPDATE, (event: FakeEvent) => this.dispatchEvent(event));
+  //   this._eventManager.listen(this._mediaSourceAdapter, Html5EventType.PLAYING, (event: FakeEvent) => this.dispatchEvent(event));
+  //   this._eventManager.listen(this._mediaSourceAdapter, Html5EventType.WAITING, (event: FakeEvent) => this.dispatchEvent(event));
+  //   this._eventManager.listen(this._mediaSourceAdapter, CustomEventType.MEDIA_RECOVERED, (event: FakeEvent) => this.dispatchEvent(event));
+  // }
+  // }
```
Comment (odedhutzler) [inline src/youtube.js]: remove?

File: src/youtube.js [hunk 11]
```diff
@@ -0,0 +1,1092 @@
...
+
+  /**
+   * Handles errors from the video element
+   * @returns {void}
+   * @private
+   */
+  _handleVideoError(): void {
+    this.dispatchEvent(new FakeEvent(EventType.ERROR, {}));
```
Comment (odedhutzler) [inline src/youtube.js]: this function is not used anywhere... what about the error handling here?

File: src/youtube.js [hunk 12]
```diff
@@ -0,0 +1,1092 @@
...
+  detach(): void {
+    Object.keys(EventType).forEach(html5Event => {
+      this._eventManager.unlisten(this._el, EventType[html5Event]);
+    });
+    // if (this._mediaSourceAdapter) {
+    //   this._eventManager.unlisten(this._mediaSourceAdapter, CustomEventType.VIDEO_TRACK_CHANGED);
+    //   this._eventManager.unlisten(this._mediaSourceAdapter, CustomEventType.AUDIO_TRACK_CHANGED);
+    //   this._eventManager.unlisten(this._mediaSourceAdapter, CustomEventType.TEXT_TRACK_CHANGED);
+    //   this._eventManager.unlisten(this._mediaSourceAdapter, CustomEventType.TEXT_CUE_CHANGED);
+    // }
```
Comment (odedhutzler) [inline src/youtube.js]: remove i think

File: src/youtube.js [hunk 13]
```diff
@@ -0,0 +1,1092 @@
...
+
+  /**
+   * Listen to the video element events and triggers them from the engine.
+   * @public
+   * @returns {void}
+   */
+  // attach(): void {
+  // Object.keys(Html5EventType).forEach(html5Event => {
+  //   this._eventManager.listen(this._el, Html5EventType[html5Event], () => {
+  //     if (Html5EventType[html5Event] === Html5EventType.ERROR) {
```
Comment (odedhutzler) [inline src/youtube.js]: why remove the listeners to the video element events?

File: src/youtube.js [hunk 14]
```diff
@@ -0,0 +1,1092 @@
...
+  _parseVideoTracks(): Array<VideoTrack>{
+    const levels: Array<string> = this._api.getAvailableQualityLevels();
+    const currentLevel: string = this._api.getPlaybackQuality();
+    const videoTracks = [];
+    levels.forEach((level, i) => {
+      let settings = {
+        active: currentLevel === level,
+        label: level,
+        bandwidth: 0,
+        width: i+1, //TODO: get estimated width and height from https://developers.google.com/youtube/iframe_api_reference#Playback_quality
```
Comment (odedhutzler) [inline src/youtube.js]: i don't understand this, i is the index in the getAvailableQualityLevels array, right?

Comment (OrenMe) [reply src/youtube.js]: For now this method is a placeholder as it seems the `setQualityLevel` method doesn't even work.
If we will have it working then we can take the suggested height/width values from the linked URL, where youtube has some outline of quality name to width/height values.

File: src/youtube.js [hunk 15]
```diff
@@ -0,0 +1,1104 @@
...
+ * Youtube engine for playback.
+ * @classdesc
+ */
+class Youtube extends FakeEventTarget implements IEngine {
+  /**
+   * The video element.
+   * @type {HTMLVideoElement}
+   * @private
+   */
+  _el: Object;
```
Comment (odedhutzler) [inline src/youtube.js]: change to HTMLVideoElement

File: src/youtube.js [hunk 16]
```diff
@@ -0,0 +1,1104 @@
...
+   */
+  static id: string = 'youtube';
+
+  /**
+   * A video element for browsers which block auto play.
+   * @type {HTMLVideoElement}
+   * @private
+   * @static
+   */
+  static _el: HTMLElement;
```
Comment (odedhutzler) [inline src/youtube.js]: also change here

File: src/youtube.js (hunk 11 above)
Comment (odedhutzler) [reply src/youtube.js]: remove this :)

File: src/youtube.js [hunk 17]
```diff
@@ -0,0 +1,1104 @@
...
+  /**
+   * Select a new video track.
+   * @param {VideoTrack} videoTrack - The video track object to set.
+   * @returns {void}
+   */
+  selectVideoTrack(videoTrack: VideoTrack): void {
+    if (this._api) {
+      if (videoTrack instanceof VideoTrack && (!videoTrack.active || this.isAdaptiveBitrateEnabled()) ) {
+        if (this.isAdaptiveBitrateEnabled()) {
+          this.dispatchEvent(EventType.ABR_MODE_CHANGED, {mode: 'manual'});
```
Comment (odedhutzler) [inline src/youtube.js]: not sure if this should be exported to a const / enum

File: src/youtube.js [hunk 18]
```diff
@@ -0,0 +1,1104 @@
...
+    this._stopSeekTargetWatchDog();
+    this._seekTargetIntervalId = setInterval(() => {
+      if (!this.paused || !this._isSeeking) {
+        this._stopSeekTargetWatchDog();
+      } else if (this.currentTime !== previousTime) {
+        this.dispatchEvent(new FakeEvent(EventType.TIME_UPDATE));
+        this.dispatchEvent(new FakeEvent(EventType.SEEKED));
+        this._isSeeking = false;
+      }
+    }, 250);
```
Comment (odedhutzler) [inline src/youtube.js]: maybe const instead of this number?

File: src/youtube.js [hunk 19]
```diff
@@ -0,0 +1,1104 @@
...
+  get ended(): boolean {
+    return this._api ? this._api.getPlayerState() === window.YT.PlayerState.ENDED : false;
+  }
+
+  /**
+   * The error property returns a MediaError object.
+   * @returns {MediaError} - The MediaError object has a code property containing the error state of the audio/video.
+   * @public
+   */
+  get error(): ?MediaError {
```
Comment (odedhutzler) [inline src/youtube.js]: where is it used?

File: src/youtube.js [hunk 20]
```diff
@@ -0,0 +1,1104 @@
...
+    return null;
+  }
+
+  /**
+   * @returns {Number} - The current network state (activity) of the audio/video.
+   * @public
+   */
+  get networkState(): number {
+    if (!(this._api && this._api.getPlayerState)) {
+      return 0;
```
Comment (odedhutzler) [inline src/youtube.js]: do we have enum for the numbers?

File: src/youtube.js [hunk 21]
```diff
@@ -0,0 +1,1104 @@
...
+    this._loadYouTubeIframeAPI().then(() => {
+      this._loadYouTubePlayer();
+    }).catch(e => {
+      dispatchError(e);
+    });
+  }
+
+  _getYoutubeErrorCodeData(errorNumber: number): {code: number, message: string} {
+    switch (errorNumber) {
+      case 2:
```
Comment (odedhutzler) [inline src/youtube.js]: not sure if we should have enum for the error numbers from youtube as well

File: src/youtube.js [hunk 22]
```diff
@@ -0,0 +1,1104 @@
...
+    this.dispatchEvent(new FakeEvent(EventType.ENDED));
+  }
+
+  _startPlayingWatchDog() {
+    this._stopPlayingWatchDog();
+    this._playingIntervalId = setInterval(() => {
+      if (!this._isSeeking) {
+        this.dispatchEvent(new FakeEvent(EventType.TIME_UPDATE));
+      }
+    }, 250);
```
Comment (odedhutzler) [inline src/youtube.js]: a const here as well

File: src/youtube.js (hunk 15 above)
Comment (OrenMe) [reply src/youtube.js]: it is actually an `HTMLElement`

File: src/youtube.js (hunk 16 above)
Comment (OrenMe) [reply src/youtube.js]: not needed - removed it

File: src/youtube.js (hunk 18 above)
Comment (OrenMe) [reply src/youtube.js]: done

File: src/youtube.js (hunk 19 above)
Comment (OrenMe) [reply src/youtube.js]: it's on the interface so I won't remove it for now

File: src/youtube.js (hunk 20 above)
Comment (OrenMe) [reply src/youtube.js]: no, and this is currently not exposed so we can add it as a separate task

File: src/youtube.js (hunk 21 above)
Comment (OrenMe) [reply src/youtube.js]: they are not enumed by Youtube so didn't add them as our enums as well

File: src/youtube.js (hunk 22 above)
Comment (OrenMe) [reply src/youtube.js]: done

File: src/youtube.js (hunk 17 above)
Comment (OrenMe) [reply src/youtube.js]: Good point. Added task to next sprint: FEC-8832 this change needs to be done across adapters and engines

File: src/youtube.js [hunk 23]
```diff
@@ -0,0 +1,1094 @@
...
+
+  /**
+   * Seeking to live edge.
+   * @function seekToLiveEdge
+   * @returns {void}
+   * @public
+   */
+  seekToLiveEdge(): void {
+    if (this._api) {
+      this._api.seekTo(this._api.getDuration()); //TODO: need to check this
```
[the remark this answers is not in the document]
Comment (yairans) [reply src/youtube.js]: if live is unsupported just leave it empty

File: src/youtube.js [hunk 24]
```diff
@@ -0,0 +1,1094 @@
...
+    }
+  }
+
+  /**
+   * Get the start time of DVR window in live playback in seconds.
+   * @returns {Number} - start time of DVR window.
+   * @public
+   */
+  getStartTimeOfDvrWindow(): number {
+    return this._api ? this._api.getDuration() : 0; //TODO: need to fix this
```
[the remark this answers is not in the document]
Comment (yairans) [reply src/youtube.js]: if live is unsupported just leave it empty

File: src/youtube.js [hunk 25]
```diff
@@ -0,0 +1,1094 @@
...
+   * @returns {Promise<Object>} - The loaded data
+   */
+  load(startTime: ?number): Promise<Object> {
+    this._loaded = true;
+    this._sdkLoaded
+      .then(() => {
+        //TODO: if autoplay pass it here and check in play if already playing?
+        const loadOptions = {
+          videoId: this._source.url,
+          startSeconds: (startTime && startTime > 0) ? startTime : 0
```
Comment (yairans) [inline src/youtube.js]:
```suggestion
          startSeconds: (startTime > 0) ? startTime : 0
```

File: src/youtube.js [hunk 26]
```diff
@@ -0,0 +1,1094 @@
...
+  _parseVideoTracks(): Array<VideoTrack>{
+    const levels: Array<string> = this._api.getAvailableQualityLevels();
+    const currentLevel: string = this._api.getPlaybackQuality();
+    const videoTracks = [];
+    levels.forEach((level, i) => {
+      let settings = {
+        active: currentLevel === level,
+        label: level,
+        bandwidth: 0,
+        width: i+1, //TODO: get estimated width and height from https://developers.google.com/youtube/iframe_api_reference#Playback_quality
```
Comment (yairans) [inline src/youtube.js]: if no value better to set to `0` or `-1` like `videoHeight` below

File: src/youtube.js [hunk 27]
```diff
@@ -0,0 +1,1094 @@
...
+   * @param {Number} to - The number to set in seconds.
+   * @public
+   * @returns {void}
+   */
+  set currentTime(to: number): void {
+    if (this._api) {
+      if (this.currentTime !== to) {
+        this._isSeeking = true;
+        this._api.seekTo(to, true);
+        this.dispatchEvent(new FakeEvent(EventType.SEEKING));
```
Comment (yairans) [inline src/youtube.js]: what's about `SEEKED`? triggered from the api?

File: src/youtube.js [hunk 28]
```diff
@@ -0,0 +1,1094 @@
...
+  ready() {}
+
+  /**
+   * Get paused state.
+   * @returns {boolean} - The paused value of the video element.
+   * @public
+   */
+  get paused(): boolean {
+    if (this._api && this._api.getPlayerState) {
+      return ![window.YT.PlayerState.PLAYING, window.YT.PlayerState.BUFFERING].includes(this._api.getPlayerState());
```
Comment (yairans) [inline src/youtube.js]: YT is nullable no?

File: src/youtube.js [hunk 29]
```diff
@@ -0,0 +1,1094 @@
...
+    return 1;
+  }
+
+  /**
+   * The ended property returns whether the playback of the audio/video has ended.
+   * @returns {boolean} - The ended value.
+   * @public
+   */
+  get ended(): boolean {
+    return this._api ? this._api.getPlayerState() === window.YT.PlayerState.ENDED : false;
```
Comment (yairans) [inline src/youtube.js]: YT is nullable no?

File: src/youtube.js [hunk 30]
```diff
@@ -0,0 +1,1094 @@
...
+
+  /**
+   * @returns {Number} - The current network state (activity) of the audio/video.
+   * @public
+   */
+  get networkState(): number {
+    if (!(this._api && this._api.getPlayerState)) {
+      return 0;
+    }
+    const playerState = window.YT.PlayerState;
```
Comment (yairans) [inline src/youtube.js]: YT is nullable no?

File: src/youtube.js [hunk 31]
```diff
@@ -0,0 +1,1094 @@
...
+  static runCapabilities(): void {}
+
+  /**
+   * Gets the youtube capabilities.
+   * @return {Promise<Object>} - The Youtube capabilities object.
+   * @public
+   * @static
+   */
+  static getCapabilities(): Promise<Object> {
+    return Promise.resolve({[Youtube.id]: {autoplay: true, mutedAutoPlay: true}});
```
[the remark this answers is not in the document]
Comment (OrenMe) [reply src/youtube.js]: Issue is that you can't know - there are no API to know this.
Our best option is to assume as YT is popular :-)

File: src/youtube.js [hunk 32]
```diff
@@ -0,0 +1,1094 @@
...
+
+  /**
+   * Select a new video track.
+   * @param {VideoTrack} videoTrack - The video track object to set.
+   * @returns {void}
+   */
+  selectVideoTrack(videoTrack: VideoTrack): void {
+    if (this._api) {
+      if (videoTrack instanceof VideoTrack && (!videoTrack.active || this.isAdaptiveBitrateEnabled()) ) {
+        if (this.isAdaptiveBitrateEnabled()) {
```
[the remark this answers is not in the document]
Comment (OrenMe) [reply src/youtube.js]: yes, cause if this is second time user selects a manual bitrate then you would still get in this method, but you won't want to dispatch `ABR_MODE_CHANGED` with manual mode

File: src/youtube.js (hunk 27 above)
Comment (OrenMe) [reply src/youtube.js]: No, they don't handle seeking or seeked at all

File: src/youtube.js (hunk 28 above)
Comment (OrenMe) [reply src/youtube.js]: If `_this._api` exist it means Iframe API was loaded

**Pull request merged**